### PR TITLE
Li/no_std support: hashmap to btreeset

### PR DESCRIPTION
**Pull request by lana-shanghai**
#BugFix

`HashMap` can be unpredictable when we encode/decode it within the Substrate runtime, so to always get the same order and result, we convert it to `BTreeMap` and then encode. 

Solves 1 out of 4 in [issue](https://github.com/keep-starknet-strange/madara/issues/907)

### CODE REVIEW

File: crates/blockifier/src/execution/contract_class.rs [hunk 1]
```diff
@@ -432,6 +434,13 @@ fn convert_entry_points_v1(
         .collect()
 }
 
+#[cfg(feature = "parity-scale-codec")]
+fn hashmap_to_btree(
+    hashmap: HashMap<EntryPointType, Vec<EntryPoint>>,
+) -> BTreeMap<EntryPointType, BTreeSet<EntryPoint>> {
```
Comment (tdelabro) [inline crates/blockifier/src/execution/contract_class.rs:440]: I think you can keep it stored as a Vec. No need to make it a BTreeSet internaly

File: crates/blockifier/src/execution/contract_class.rs [hunk 2]
```diff
@@ -263,10 +266,9 @@ pub struct ContractClassV1Inner {
 impl Encode for ContractClassV1Inner {
     fn encode(&self) -> Vec<u8> {
         let val = self.clone();
-        let entry_points_by_type = val
-            .entry_points_by_type
-            .into_iter()
-            .collect::<Vec<(EntryPointType, Vec<EntryPointV1>)>>();
+        let entry_point_btree = hashmap_to_btree(val.entry_points_by_type);
+        let entry_points_by_type =
+            entry_point_btree.into_iter().collect::<Vec<(EntryPointType, Vec<EntryPointV1>)>>();
```
Comment (tdelabro) [inline crates/blockifier/src/execution/contract_class.rs:271]: I don't think you need to do hashmap -> btree -> vec -> encode, you can do hashmap -> btree -> encode

**Review by tdelabro — CHANGES_REQUESTED** [review]
Is there no `Decode` implementation?

File: crates/blockifier/src/execution/contract_class.rs [hunk 3]
```diff
@@ -432,6 +431,20 @@ fn convert_entry_points_v1(
...
+    hashmap.into_iter().collect()
+}
+
+#[cfg(feature = "parity-scale-codec")]
+fn hashmap_to_btree_v1(
+    hashmap: HashMap<EntryPointType, Vec<EntryPointV1>>,
+) -> BTreeMap<EntryPointType, Vec<EntryPointV1>> {
+    hashmap.into_iter().collect()
+}
+
```
Comment (tdelabro) [inline crates/blockifier/src/execution/contract_class.rs:447]: Don't create two functions for this. Just write `hashmap.into_iter().collect()` directly in the original function body

File: crates/blockifier/src/execution/contract_class.rs [hunk 4]
```diff
@@ -149,8 +149,9 @@ impl Decode for ContractClassV0Inner {
         input: &mut I,
     ) -> Result<Self, parity_scale_codec::Error> {
         let res = <(Program, Vec<(EntryPointType, Vec<EntryPoint>)>)>::decode(input)?;
+        let entry_point_btree = <BTreeMap<EntryPointType, Vec<EntryPoint>>>::from_iter(res.1);
         let entry_points_by_type =
-            <HashMap<EntryPointType, Vec<EntryPoint>>>::from_iter(res.1.into_iter());
+            <HashMap<EntryPointType, Vec<EntryPoint>>>::from_iter(entry_point_btree);
```
Comment (tdelabro) [inline crates/blockifier/src/execution/contract_class.rs:154]: Not sure about this. If you converted it to a btree before encoding, decode it as a btree too. Not as as `Vec<(EntryPointType, Vec<EntryPoint>)>`.
And then you just have to iterate once and collect into a `HashMap`

File: crates/blockifier/src/execution/contract_class.rs [hunk 5]
```diff
@@ -281,9 +279,10 @@ impl Decode for ContractClassV1Inner {
             <(Program, Vec<(EntryPointType, Vec<EntryPointV1>)>, Vec<(String, Hint)>)>::decode(
                 input,
             )?;
+        let entry_point_btree = <BTreeMap<EntryPointType, Vec<EntryPointV1>>>::from_iter(res.1);
         let entry_points_by_type =
-            <HashMap<EntryPointType, Vec<EntryPointV1>>>::from_iter(res.1.into_iter());
-        let hints = <HashMap<String, Hint>>::from_iter(res.2.into_iter());
+            <HashMap<EntryPointType, Vec<EntryPointV1>>>::from_iter(entry_point_btree);
+        let hints = <HashMap<String, Hint>>::from_iter(res.2);
```
Comment (tdelabro) [inline crates/blockifier/src/execution/contract_class.rs:285]: Same here. Decode to Btree and iter only once

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/blockifier/src/execution/contract_class.rs [hunk 6]
```diff
@@ -157,8 +160,9 @@ impl Decode for ContractClassV0Inner {
         input: &mut I,
     ) -> Result<Self, parity_scale_codec::Error> {
         let res = <(Program, Vec<(EntryPointType, Vec<EntryPoint>)>)>::decode(input)?;
+        let entry_point_btree = <BTreeMap<EntryPointType, Vec<EntryPoint>>>::from_iter(res.1);
```
Comment (tdelabro) [inline crates/blockifier/src/execution/contract_class.rs:163]: This is not needed.
On l162 decode into a `<BTreeMap<EntryPointType, Vec<EntryPoint>>>` directly. And then you only have to iterate once in order to collect into the final hashmap

**Review by tdelabro — CHANGES_REQUESTED** [review]

**Comment by shramee** [comment]
Hi guys,
This PR removes `spin` but it's still in use in `blockifier::sync`.

Perhaps we should test with `--no-default-features` in CI?

image

**Pull request merged**